### ISSUE TITLE
cilium: ipv6 built-in in kernel, unable to flush Cilium ip{6,}tables chain

### ISSUE THREAD

**Issue by borkmann** [cilium/cilium#10134]
Seen in 1.6.6 with `--enable-ipv6=true` option.

```
NAME="Ubuntu"
VERSION="18.04.4 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.4 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

IPv6 is built-in:

```
vagrant@vagrant:~/cilium/examples/getting-started$ modprobe --show-depends ipv6
builtin ipv6
```

```
cilium-docker-plugin | level=info msg="Waiting for cilium daemon to start up..." subsys=cilium-docker-driver url="unix:///var/run/cilium/cilium.sock"
cilium           | level=warning msg="Unable to flush Cilium ip6tables chain" error="exit status 3" obj="[-w 5 -t mangle -F CILIUM_PRE_mangle]" subsys=iptables
cilium           | level=warning msg="Unable to delete Cilium ip6tables chain" error="exit status 3" obj="[-w 5 -t mangle -X CILIUM_PRE_mangle]" subsys=iptables
cilium           | level=warning msg="Unable to flush Cilium iptables chain" error="exit status 1" obj="[-w 5 -t raw -F CILIUM_PRE_raw]" subsys=iptables
cilium           | level=warning msg="Unable to delete Cilium iptables chain" error="exit status 1" obj="[-w 5 -t raw -X CILIUM_PRE_raw]" subsys=iptables
cilium           | level=warning msg="Unable to flush Cilium ip6tables chain" error="exit status 3" obj="[-w 5 -t raw -F CILIUM_PRE_raw]" subsys=iptables
cilium           | level=warning msg="Unable to delete Cilium ip6tables chain" error="exit status 3" obj="[-w 5 -t raw -X CILIUM_PRE_raw]" subsys=iptables
cilium           | level=warning msg="Unable to flush Cilium iptables chain" error="exit status 1" obj="[-w 5 -t filter -F CILIUM_FORWARD]" subsys=iptables
cilium           | level=warning msg="Unable to delete Cilium iptables chain" error="exit status 1" obj="[-w 5 -t filter -X CILIUM_FORWARD]" subsys=iptables
cilium           | level=error msg="Command execution failed" cmd="[ip6tables -w 5 -t filter -N CILIUM_INPUT]" error="exit status 3" subsys=iptables
cilium           | level=warning msg="modprobe: ERROR: ../libkmod/libkmod.c:586 kmod_search_moddep() could not open moddep file '/lib/modules/4.15.0-76-generic/modules.dep.bin'" subsys=iptables
cilium           | level=warning msg="modprobe: FATAL: Module ip6_tables not found in directory /lib/modules/4.15.0-76-generic" subsys=iptables
cilium           | level=warning msg="ip6tables v1.6.1: can't initialize ip6tables table `filter': Table does not exist (do you need to insmod?)" subsys=iptables
cilium           | level=warning msg="Perhaps ip6tables or your kernel needs to be upgraded." subsys=iptables
cilium           | level=error msg="Error while initializing daemon" error="cannot add custom chain CILIUM_INPUT: exit status 3" subsys=daemon
cilium           | level=fatal msg="Error while creating daemon" error="cannot add custom chain CILIUM_INPUT: exit status 3" subsys=daemon
```

```
cilium           | level=warning msg="IPv6 is enabled and ip6tables modules could not be initialized" error="could not load module ip6_tables: exit status 1" subsys=iptables                                        
cilium           | level=warning msg="xt_socket kernel module could not be loaded" error="could not load module xt_socket: exit status 1" subsys=iptables   
```

**Comment by borkmann** [comment]
Could not repro so far.

**Comment by olemarkus** [comment]
I get all these errors on 1.6.6 as well. I run ipv6 disabled, so I only get the iptables errors though.

**Comment by joestringer** [comment]
Might be worth trying this in a GKE environment?

**Comment by errordeveloper** [comment]
This looks like it should be revisited. What's not clear from the issue as it stand is what's the impact of this? Generally, it sounds like Cilium makes has some functionality that have a hard assumption on certain kernel features being a module, is that correct? In which case, it's probably a good idea to enhance the checks to cater for builtins.

**Comment by joestringer** [comment]
Impact looks like Cilium crashes with a fatal warning and refuses to start.

FWIW, I tried running Cilium 1.8.5 today with `enable-ipv6: true`, `cluster-pool-ipv6-cidr: fd00::/48` and `cluster-pool-ipv6-mask-size: "64"` under microk8s on Ubuntu 20.04 (kernel 5.8 vanilla) and I saw no such warnings. IPv6 is built in just like was reported in the issue description.

I would welcome reports from anyone who is having problems with this and how to reproduce. If we haven't heard anything in six months then it's not obvious to me that any users are actively facing problems here.
